### PR TITLE
Issues/15 running games in parallel

### PR DESCRIPTION
**Pull request by donren-leung**
## Main
Multiprocessing! Games in each round are run across a pool of CPUs.

```
$ poetry run checkers RandomBot FirstMover CopyCat ScaredyCat GreedyCat  --mode all --rounds 4 --verbose --output-dir output
5 bots registered
20 double-round-robin games/tourney * 4 tourneys = 80 games scheduled
Round 0 completed in 0.5 seconds
Round 1 completed in 0.5 seconds
Round 2 completed in 0.5 seconds
Round 3 completed in 0.5 seconds
```

## Others
- One subfolders for each round in output folder and each game file has both player names, for better organisation and findability.
- Bot get_name is class method.
- BotTracker doesn't possess any Bot instances, but you can call spawn_bot like in Game.run(). Subfolders for rounds in output.

### CODE REVIEW

File: checkers_bot_tournament/game.py [hunk 1]
```diff
@@ -15,19 +16,18 @@
...
         start_pdn: Optional[str],
     ):
-        self.white = white
-        self.black = black
-
-        self.white.reset_bot()
-        self.black.reset_bot()
+        self.white_tracker = white_tracker
+        self.black_tracker = black_tracker
+        self.white_bot: Bot | None = None
```
Comment (trentzz) [inline checkers_bot_tournament/game.py]: These should be `Optional[Bot]`

File: checkers_bot_tournament/controller.py [hunk 2]
```diff
@@ -92,6 +96,8 @@ def __init__(
         self.game_results: list[list[GameResult]] = [[] for _ in range(rounds)]
         self.game_id_counter: int = 0
         self.game_results_folder: Optional[str] = None
+        self.write_queue: Queue[tuple[list[GameResult], int] | None] = Queue()
```
Comment (trentzz) [inline checkers_bot_tournament/controller.py]: Should also be `Optional[tuple[list[GameResult], int]]`

File: checkers_bot_tournament/controller.py [hunk 3]
```diff
@@ -219,50 +225,87 @@ def _create_timestamped_folder(self, prefix: str = "checkers_game_results") -> N
...
     def _write_game_result_summary(self, file: IO, game_result: GameResult) -> None:
         file.write(str(game_result))
         file.write("\n" + "=" * 40 + "\n")
 
-    def _write_game_results(self, game_results: list[GameResult]) -> None:
+    def _write_game_results(self, game_results: list[GameResult], round_number: int) -> None:
         assert self.game_results_folder is not None
+        Path: TypeAlias = str
+
+        round_folder_suffix: Path = f"round_{round_number}"
```
Comment (trentzz) [inline checkers_bot_tournament/controller.py]: This makes `round_X` folders with nothing in them even when `--verbose` is not specified btw, should check if `self.verbose` first

File: checkers_bot_tournament/controller.py (hunk 2 above)
Comment (donren-leung) [reply checkers_bot_tournament/controller.py]: `Optional` and `|` are equivalent so it's a matter of style. 
https://peps.python.org/pep-0604/
https://stackoverflow.com/questions/69440494/python-3-10-optionaltype-or-type-none

Especially here I was thinking this was already way too nested to add another layer.

Comment (trentzz) [reply checkers_bot_tournament/controller.py]: Yeah it is defs a style thing, I'd rather keep it all consistent (i.e. either we use `| None` for everything, or `Optional` for everything and I much much prefer `Optional`), nested is totally fine imo but yea nitpicky

Comment (trentzz) [reply checkers_bot_tournament/controller.py]: Maybe my style preference is also coming from all the other languages (cpp/rust/scala/etc) that all use `Optional`

**Pull request merged**